MericCelik/assignment3/TrieBasedDictionary.java

```diff
@@ -0,0 +1,140 @@
+     * @return set of possible words(no duplicates)
+     * @throws NullPointerException when the parameters are invalid.
+     *
+     * The logic is the same with the Main's but it uses TrieBasedDictionary, so when it calls buildAllWords() method,
+     * isWord() and isPrefix() method is much faster.
+     * */
+    public static HashSet<String> findWords(String[][] grids, TrieBasedDictionary dict){
+
+        if(grids == null || dict ==null)
+            throw new NullPointerException("Invalid parameter!");
+
+        int rowcount = grids.length;
+        int columncount = grids[0].length;
+
+        HashSet<String> allresults = new HashSet<>();
+        ArrayList<String> result;
+        for(int r=0;r<rowcount;r++){
+            for(int c=0;c<columncount;c++){
+                int[][] visited = new int[rowcount][columncount];
+                result = buildAllWords(grids,r,c,"",visited,dict, new ArrayList<>());
+                allresults.addAll(result);
+            }
+        }
+        return allresults;
+    }
+
+    /** Builds all possible words that begins with given character. It is a recursive algorithm, so possible words are
+     * stored in the parameter named "result". Current word is stores in the parameter "word".
+     * The method adds the character to the word and check whether the dictionary contains this word as a prefix. If not,
+     * there is no need to add new letters from this word, so it returns current result array. If contains, method continues
+     * and check whether the dictionary contains this word as a complete word. If not, it does not add it to the result array.
+     * After that, it check all neighbour grids and call the same function. During this process, it
+     * updates the visited array for each possible words. A word cannot use the same grid twice, so it keeps track of it.
+     * Complexities: Best Case: O(M*N) when the dictionary has none of the grids as a prefix.
+     *              Worst Case: O((M*N)^2) when the dictionary has all the words can be constructed with grids.
+     *              Average Case: O((M*N)^2), M and N are row and column counts.
+     * @param grids : The grids matrix to construct words from its elements.
+     * @param dict : The dictionary that contains some words and prefixes.
+     * @param r : current row coordinate/number
+     * @param c: current column coordinate/number
+     * @param word : current word it is constructing
+     * @param result : possible words it finds so far
+     * @param visited : binary array to keep track what grids are visited while constructing current word.
+     * @return list of possible words (no duplicates)
+     *
+     * The logic is the same with the Main's but it uses TrieBasedDictionary, so isWord() and isPrefix() method is much
+     * faster.
+     * */
+    public static ArrayList<String> buildAllWords(String[][] grids, int r, int c, String word, int[][] visited,
```

Comment:
Seems like a duplication from the main class. Maybe have TrieBasedDictionary either extend Dictionary or have both implement BasicDictionary interface, so you can avoid this code duplication?

Reply:
Merged two functions. :+1: I added one extremely simple interface too! :smile: It helped me to remember the usage of it, thank you!